IdentityCore/src/cache/mac/MSIDMacSharedCredentialCacheItem.m

```diff
@@ -0,0 +1,190 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDMacSharedCredentialCacheItem.h"
+
+static NSString *keyDelimiter = @"-";
+
+@interface MSIDMacSharedCredentialCacheItem ()
+
+@property (nonatomic) NSMutableDictionary *cacheObjects;
+@property (nonatomic) dispatch_queue_t queue;
+
+@end
+
+@implementation MSIDMacSharedCredentialCacheItem
+
+- (instancetype)initPrivate
+{
+    if(self = [super init])
+    {
+        self.cacheObjects = [NSMutableDictionary dictionary];
+        self.queue = dispatch_queue_create("com.microsoft.universalStorage", DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return self;
+}
+
++ (MSIDMacSharedCredentialCacheItem *)sharedInstance
+{
+    static MSIDMacSharedCredentialCacheItem *instance = nil;
+    
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        instance = [[self alloc] initPrivate];
+    });
+    
+    return instance;
+}
+
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json
+                                 error:(NSError * __autoreleasing *)error
+{
+    if (!(self = [self initPrivate]))
+    {
+        return nil;
+    }
+    
+    for (NSString *refreshTokenKey in json)
+    {
+        NSDictionary *rtDict = [json objectForKey:refreshTokenKey];
+        
+        if (rtDict)
+        {
+            MSIDCredentialCacheItem *refreshToken = [[MSIDCredentialCacheItem alloc] initWithJSONDictionary:rtDict error:error];
+            
+            if (refreshToken)
+            {
+                MSIDDefaultCredentialCacheKey *key = [[MSIDDefaultCredentialCacheKey alloc] initWithHomeAccountId:refreshToken.homeAccountId                      environment:refreshToken.environment clientId:refreshToken.clientId credentialType:refreshToken.credentialType];
+                
+                key.familyId = refreshToken.familyId;
+                key.realm = refreshToken.realm;
+                key.target = refreshToken.target;
+                key.enrollmentId = refreshToken.enrollmentId;
+                
+                [self setSharedCredential:refreshToken forKey:key];
+            }
+        }
+    }
+    
+    return self;
+}
+
+- (NSDictionary *)jsonDictionary
+{
+    NSDictionary *cacheObjects = [self getCopy];
+    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+    
+    for (MSIDDefaultCredentialCacheKey *refreshTokenKey in cacheObjects)
```

Comment:
Instead of copying cacheObjects, dispatch calls wherever it's used should be used to protect it. Here, I think a dispatch_sync() would be appropriate.

Reply:
Lets discuss this offline as I see a potential issue with this. I am already using dispatch_sync in the copy method but a deep copy is important for some reasons.